src/core/http1/client.rs

```diff
@@ -0,0 +1,604 @@
+/*
+ * Copyright (C) 2020-2023 Fanout, Inc.
+ * Copyright (C) 2023-2024 Fastly, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use crate::buffer::{Buffer, VecRingBuffer, VECTORED_MAX};
+use crate::core::http1::error::Error;
+use crate::core::http1::util::*;
+use crate::future::{
+    select_2, AsyncRead, AsyncWrite, AsyncWriteExt, ReadHalf, Select2, StdWriteWrapper, WriteHalf,
+};
+use crate::http1;
+use crate::pin;
+use std::cell::RefCell;
+use std::io::{self, Write};
+use std::mem;
+use std::pin::Pin;
+use std::str;
+
+pub struct Request<'a, R: AsyncRead, W: AsyncWrite> {
+    r: ReadHalf<'a, R>,
+    w: WriteHalf<'a, W>,
+    buf1: &'a mut VecRingBuffer,
+    buf2: &'a mut VecRingBuffer,
+}
+
+impl<'a, R: AsyncRead, W: AsyncWrite> Request<'a, R, W> {
+    pub fn new(
+        stream: (ReadHalf<'a, R>, WriteHalf<'a, W>),
+        buf1: &'a mut VecRingBuffer,
+        buf2: &'a mut VecRingBuffer,
+    ) -> Self {
+        Self {
+            r: stream.0,
+            w: stream.1,
+            buf1,
+            buf2,
+        }
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub fn prepare_header(
+        self,
+        method: &str,
+        uri: &str,
+        headers: &[http1::Header<'_>],
+        body_size: http1::BodySize,
+        websocket: bool,
+        initial_body: &[u8],
+        end: bool,
+    ) -> Result<RequestHeader<'a, R, W>, Error> {
+        let req = http1::ClientRequest::new();
+
+        let req_body = match req.send_header(self.buf1, method, uri, headers, body_size, websocket)
+        {
+            Ok(ret) => ret,
+            Err(_) => return Err(Error::BufferExceeded),
+        };
+
+        if self.buf2.write_all(initial_body).is_err() {
+            return Err(Error::BufferExceeded);
+        }
+
+        Ok(RequestHeader {
+            r: self.r,
+            w: self.w,
+            buf1: self.buf1,
+            buf2: self.buf2,
+            req_body,
+            end,
+        })
+    }
+}
+
+pub struct RequestHeader<'a, R: AsyncRead, W: AsyncWrite> {
+    r: ReadHalf<'a, R>,
+    w: WriteHalf<'a, W>,
+    buf1: &'a mut VecRingBuffer,
+    buf2: &'a mut VecRingBuffer,
+    req_body: http1::ClientRequestBody,
+    end: bool,
+}
+
+impl<'a, R: AsyncRead, W: AsyncWrite> RequestHeader<'a, R, W> {
+    pub async fn send(mut self) -> Result<RequestBody<'a, R, W>, Error> {
```

Comment:
if an error occurs during the write, the buffer's read state is not rolled back, which could lead to inconsistencies. would it make sense to implement a transaction-like mechanism where buffer states can be rolled back to maintain consistency in case of errors?

Reply:
it's okay here since the operation can only be attempted once (`send()` consumes self)